opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -304,23 +316,33 @@ def add_link(
         link_target_context: "trace_api.SpanContext",
         attributes: types.Attributes = None,
     ) -> None:
-        if self.links is Span.empty_links:
-            self.links = BoundedList(MAX_NUM_LINKS)
+        with self._lock:
+            if self.end_time is not None:
+                logger.warning("Calling add_linl() on an ended span.")
+                return
+            if self.links is Span.empty_links:
+                self.links = BoundedList(MAX_NUM_LINKS)
         if attributes is None:
             attributes = Span.empty_attributes
         self.links.append(Link(link_target_context, attributes))
 
     def start(self):
-        if self.start_time is None:
-            self.start_time = util.time_ns()
-            self.span_processor.on_start(self)
+        with self._lock:
+            if self.start_time is None:
+                self.start_time = util.time_ns()
+                self.span_processor.on_start(self)
 
     def end(self):
-        if self.end_time is None:
-            self.end_time = util.time_ns()
-            self.span_processor.on_end(self)
+        with self._lock:
+            if self.end_time is None:
+                self.end_time = util.time_ns()
+                self.span_processor.on_end(self)
```

Comment:
Same as above, I think we only need to hold the lock until `end_time` is set.

Reply:
You're right, will fix as well as above case.